source/brailleDisplayDrivers/handyTech.py

```diff
@@ -3,130 +3,831 @@
+	"VID_1FE4&PID_0061", # Actilino
+	"VID_1FE4&PID_0064", # Active Star 40
+}
 
-	def sayString(self, text):
-		speech.speakMessage(text)
+# Some older displays use a HID converter and an internal serial interface
+USB_IDS_HID_CONVERTER = {
+	"VID_1FE4&PID_0003", # USB-HID adapter
+	"VID_1FE4&PID_0074", # Braille Star 40
+	"VID_1FE4&PID_0044", # Easy Braille
+}
 
-	def onKeysPressed(self, keys_arg, routing_pos):
-		# keys_arg is VARIANT. Indexing by 0 gives actual value.
-		keys = keys_arg[0]
-		if constants.KEY_ROUTING in keys:
-			gesture = InputGesture(keys, routing_pos -1)
-		else:
-			gesture = InputGesture(keys)
-		try:
-			inputCore.manager.executeGesture(gesture)
-		except inputCore.NoInputGestureAction:
-			pass
+USB_IDS_HID = USB_IDS_HID_NATIVE | USB_IDS_HID_CONVERTER
 
-class BrailleDisplayDriver(braille.BrailleDisplayDriver, ScriptableObject):
-	"""Handy Tech braille display driver.
+# pylint: disable=C0330
+BLUETOOTH_NAMES = {
+	"Actilino AL",
+	"Active Braille AB",
+	"Active Star AS",
+	"Basic Braille BB",
+	"Braille Star BS",
+	"Braille Wave BW",
+	"Easy Braille EBR",
+}
+
+# Model identifiers
+# pylint: disable=C0103
+MODEL_BRAILLE_WAVE = "\x05"
+MODEL_MODULAR_EVOLUTION_64 = "\x36"
+MODEL_MODULAR_EVOLUTION_88 = "\x38"
+MODEL_MODULAR_CONNECT_88 = "\x3A"
+MODEL_EASY_BRAILLE = "\x44"
+MODEL_ACTIVE_BRAILLE = "\x54"
+MODEL_CONNECT_BRAILLE_40 = "\x55"
+MODEL_ACTILINO = "\x61"
+MODEL_ACTIVE_STAR_40 = "\x64"
+MODEL_BASIC_BRAILLE_16 = "\x81"
+MODEL_BASIC_BRAILLE_20 = "\x82"
+MODEL_BASIC_BRAILLE_32 = "\x83"
+MODEL_BASIC_BRAILLE_40 = "\x84"
+MODEL_BASIC_BRAILLE_48 = "\x8A"
+MODEL_BASIC_BRAILLE_64 = "\x86"
+MODEL_BASIC_BRAILLE_80 = "\x87"
+MODEL_BASIC_BRAILLE_160 = "\x8B"
+MODEL_BRAILLE_STAR_40 = "\x74"
+MODEL_BRAILLE_STAR_80 = "\x78"
+MODEL_MODULAR_20 = "\x80"
+MODEL_MODULAR_80 = "\x88"
+MODEL_MODULAR_40 = "\x89"
+
+# Key constants
+KEY_B1 = 0x03
+KEY_B2 = 0x07
+KEY_B3 = 0x0B
+KEY_B4 = 0x0F
+KEY_B5 = 0x13
+KEY_B6 = 0x17
+KEY_B7 = 0x1B
+KEY_B8 = 0x1F
+KEY_LEFT = 0x04
+KEY_RIGHT = 0x08
+KEY_LEFT_SPACE = 0x10
+KEY_RIGHT_SPACE = 0x18
+KEY_ROUTING = 0x20
+KEY_RELEASE = 0x80
+
+# Braille dot mapping
+KEY_DOTS = {
+	KEY_B4: 1,
+	KEY_B3: 2,
+	KEY_B2: 3,
+	KEY_B5: 4,
+	KEY_B6: 5,
+	KEY_B7: 6,
+	KEY_B1: 7,
+	KEY_B8: 8,
+}
+
+# Considered spaces in braille input mode
+KEY_SPACES = (KEY_LEFT_SPACE, KEY_RIGHT_SPACE,)
+
+
+class Model(AutoPropertyObject):
+	"Extend from this base class to define model specific behavior."
+	#: Device identifier, used in the protocol to identify the device
+	#: @type: string
+	deviceId = None
+
+	#: A reference to the driver instance
+	#: @type; BrailleDisplayDriver
+	_display = None
+
+	#: A generic name that identifies the model/series, used in gesture identifiers
+	#: @type: string
+	genericName = None
+
+	#: Specific name of this model
+	#: @type: string
+	name = None
+
+	#: Number of braille cells
+	#: @type: int
+	numCells = 0
+
+	def __init__(self, display):
+		super(Model, self).__init__()
+		self._display = display

```

Comment:
Setting _display to display and _model to model causes a circular reference. If __del__ is ever implemented on either of those classes, the garbage collector will not be able to clean them up.  I don't think this is happening right now. But it may be safer to make _display a weakreference. This should be okay as model instances only ever exist within the lifetime of a display.